mteb/abstasks/mteb_result.py

```diff
@@ -0,0 +1,183 @@
+        for split, hf_subset_scores in v.items():
+            for lang, scores in hf_subset_scores.items():
+                cls._validate_scores_dict(scores)
+        return v
+
+    @staticmethod
+    def _validate_scores_dict(scores: Scores) -> None:
+        if "evaluation_time" not in scores:
+            raise ValueError("'evaluation_time' should be in scores")
+
+        if "main_score" not in scores:
+            raise ValueError("'main_score' should be in scores")
+
+        # check that it is json serializable
+        try:
+            _ = json.dumps(scores)
+        except Exception as e:
+            raise ValueError(f"Scores are not json serializable: {e}")
+
+    def to_dict(self) -> dict:
+        return self.model_dump()
+
+    @classmethod
+    def from_dict(cls, data: dict) -> MTEBResults:
+        return cls.model_validate(data)
+
+    def to_disk(self, path: Path) -> None:
+        with path.open("w") as f:
+            f.write(self.model_dump_json())
+
+    @classmethod
+    def from_disk(cls, path: Path) -> MTEBResults:
+        with path.open("r") as f:
+            return cls.model_validate(json.load(f))
+
+    def get_score(
+        self,
+        getter: Callable[[Scores], float],
+        aggregation: Callable[[list[float]], Any],
+        splits: list[Split] | None = None,
+        languages: list[ISO_LANGUAGE | ISO_LANGUAGE_SCRIPT] | None = None,
+        scripts: list[ISO_LANGUAGE_SCRIPT] | None = None,
+    ) -> Any:
+        """Get a score for the specified splits, languages, scripts and aggregation function.
+
+        Args:
+            getter: A function that takes a scores dictionary and returns a score e.g. "main_score" or "evaluation_time".
+            aggregation: The aggregation function to use.
+            splits: The splits to consider.
+            languages: The languages to consider. Can be ISO language codes or ISO language script codes.
+            scripts: The scripts to consider.
+
+        Returns:
+            The result of the aggregation function on the scores.
+        """
+        name2task = create_name_to_task_mapping()
+        meta = name2task[self.task_name].metadata
+        hf_subset2langs = eval_langs_as_dict(meta.eval_langs)
+
+        if splits is None:
+            splits = list(self.scores.keys())
+
+        lang_script_filter = LangScriptFilter.from_languages_and_scripts(
+            languages, scripts
+        )
+
+        values = []
+        for split in splits:
+            if split not in self.scores:
+                raise ValueError(f"Split {split} not found in scores")
+
+            for hf_subset, scores in self.scores[split].items():
+                eval_langs = hf_subset2langs[hf_subset]
+                for lang in eval_langs:
+                    if lang_script_filter.is_valid_language(lang):
+                        values.append(getter(scores))
+                        break
+
+            return aggregation(values)
+
+    def get_main_score(
```

Comment:
This can extract scores for a specific language, split or script. Useful for going from hf_subset to language codes.

Reply:
See example in docstring.

Reply:
Hmm I'm not sure how I feel about `getter` my feeling is that getter is something you use in programming when you have a private variable whose reference should not be exposed, but its value should be accessible by other objects.
I don't know if I could come up with a better name for this though.

Is it not the main idea though that we should somehow be able to allow users to expose an attribute of a dictionary at arbitrary depths? In that case I would prefer some path to a deeply nested value.
e.g. if you have a Scores dict like this:
```python
{
   "something": {
        "something_else": {"beautiful_score": 0.76}
    }
}
```
Then you could have a path to this score:
```python
get_score(..., score_path=("something", "something_else", "beautiful_score"))
```
Or I've seen some libraries for accessing these with a dot syntax:
```
"something.something_else.beautiful_score"
```

Reply:
Are we discussing the get_main_score or the getter argument in get_score here?

I don't see a reason why not simply using a function. the path seems to unreasonably limit the user who e.g. wants the mean of recall_at_1 and ndgc_at_10 and it doesn't require custom syntax. 

> The getter function is used to retrieve the variable value

I believe this fits with my intuition as well

Reply:
Mmm that's a good example, I accept :D
I'm a bit uncomfortable with the name `getter` but might just be my OOP trauma from earlier years, if others don't take issue with it I'll give it a pass

Reply:
If the idea is to make it simple for users to specify a metric without custom syntax, could we accept the metric directly instead, and do all the "custom syntax" ourselves? i.e. no getter, no dots. 

Results have this structure. `main_score` is at the same level as other metrics.
```
"aau_Latn-eng_Latn": {
      "accuracy": 0.0703125,
      "f1": 0.05066417191876751,
      "main_score": 0.05066417191876751,
      "precision": 0.044970703125,
      "recall": 0.0703125
    },
```

Option here is for `get_score` to accept a string and defaults to `"main_score"`. That way we don't even need `get_main_score`. Users can directly call `get_score`. Under the hoop (in the method), we handle the logic to extract the specified metric in the structure.
```
metric: str = "main_score"
```

Reply:
I have now refactored out get_main_score such that it is:

```
def get_score(
        ...
        getter: Callable[[Scores], float] = lambda scores: scores["main_score"],
        aggregation: Callable[[list[float]], Any] = np.mean,
    )

...
mteb_results.get_score() # will get main score 
```

Reply:
I really like the aggregation option. My only last suggestion is to make the `get_score` signature simpler (sorry if it wasn't clear before):
```diff
def get_score(
        ...
-       getter: Callable[[Scores], float] = lambda scores: scores["main_score"],
+       metric: str = "main_score",
        aggregation: Callable[[list[float]], Any] = np.mean,
    )
```

then in the method, we use `lambda scores: scores[metric]`. That way the user will only need to give us the metric they want, and not a function.